File: Leetcode/Russian Doll Envelopes/Russian Doll Envelopes.c
// You are given a 2D array of integers envelopes where envelopes[i] = [wi, hi] represents the width and the height of an envelope.
// One envelope can fit into another if and only if both the width and height of one envelope is greater than the width and height of the 
// other envelope.
// Return the maximum number of envelopes can you Russian doll (i.e., put one inside the other).



// Function to calculate maximum of two integer
int max(int num1, int num2)
{
    return (num1 > num2 ) ? num1 : num2;
}
int maxEnvelopes(int** envelopes, int envelopesSize, int* envelopesColSize){

	int size = envelopesSize;
    int temp = 0;  
	if(!size) return 0;
	// sort the envelopes array by width of envelope
     for (int i = 0; i < envelopesSize; i++) {     
        for (int j = i+1; j < envelopesSize; j++) {     
           if(envelopes[i][0] > envelopes[j][0]) {    
               temp = envelopes[i][0];    
               envelopes[i][0] = envelopes[j][0];    
               envelopes[j][0] = temp;    
           }     
        }     
    }    
    int* maxrolls;
    maxrolls = calloc(envelopesSize, sizeof(int) );
    int maxroll = 1;
	maxrolls[0] = 1;
    // Traversing the Envelope
	for(int i = 1; i < size; ++i)
	{
		maxrolls[i] = 1;
		for(int j = i-1; j >= 0; --j)
            // Checking if the envelope can be Russian Dolled
             // allocating an array to record the maximal amount for each envelope
			if(envelopes[i][0]>envelopes[j][0] && envelopes[i][1]>envelopes[j][1])
				maxrolls[i] = max(maxrolls[i], maxrolls[j]+1);
       // finding the max of two
		maxroll = max(maxroll, maxrolls[i]);
	}
    // the maximum number of envelopes that can be Russian dolled
	return maxroll;
}

File: Leetcode/Russian Doll Envelopes/RussianDollEnvelopes.java
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        //sorting the array on basis of first entity i.e. height of envelope
        Arrays.sort(envelopes, (a, b) -> Integer.compare(a[0], b[0]));

        //taking LIS of the sorted array on the basis of second entity i.e. width of envelope.
        //By this we can ensure we get both height and width of envelopes in strictly increasing order so as to fit in the maximum envelopes
        int[] dp = new int[envelopes.length];
        Arrays.fill(dp, 1);
        int max_=1;
        for(int i=1; i<envelopes.length; i++){
            for(int j=i-1; j>=0; j--){
                if(envelopes[j][0] < envelopes[i][0] && envelopes[j][1]< envelopes[i][1]){
                    dp[i]= Math.max(dp[i], dp[j]+1);
                }
            }
            max_= Math.max(max_, dp[i]);
        }
        return max_;
    }
}